src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -2784,8 +2784,11 @@ public virtual bool TryGetUri(object entity, out Uri identity)
                 (callback, state) =>
                 {
                     IAsyncResult asyncResult = beginMethod(callback, state);
-                    cancellationToken.Register(() => this.CancelRequest(asyncResult));
-                    cancellationToken.ThrowIfCancellationRequested();
+                    using (CancellationTokenRegistration cancellationTokenRegistration = cancellationToken.Register(() => this.CancelRequest(asyncResult)))

```

Comment:
I'm only just learning about `CancellationTokenRegistration`s, but I think this is not the desired behavior for us. We want the callback to be executed when the token is cancelled, but by disposing the registration, I think that callback is unregistered. It appears that disposing the `CancellationTokenSource` disposes all associated registrations, so I think the memory leak would be correct (without changing existing functionality) if we would call `Dispose` on all appropriate `CancellationTokenSource`s. Am I grossly misunderstanding this?

Reply:
In this case, our goal is to ensure `CancellationTokenRegistration.Unregister` is called, that's why it's important to scope the disposable `CancellationTokenRegistration` with `using`.

In these internal `FromAsync` methods,, we are passing a `CancellationToken` not a `CancellationTokenSource`.